impl/src/main/java/com/okta/sdk/impl/oauth2/AccessTokenRetrieverServiceImpl.java

```diff
@@ -0,0 +1,90 @@
+package com.okta.sdk.impl.oauth2;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.okta.commons.http.HttpException;
+import com.okta.commons.http.HttpMethod;
+import com.okta.commons.lang.Assert;
+import com.okta.commons.lang.Strings;
+import com.okta.sdk.impl.config.ClientConfiguration;
+import com.okta.sdk.oauth2.ErrorResponse;
+import com.okta.sdk.oauth2.OAuth2AccessToken;
+import okhttp3.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+
+import static com.okta.sdk.impl.util.OAuth2Utils.createSignedJWT;
+
+public class AccessTokenRetrieverServiceImpl implements AccessTokenRetrieverService {
+    private static final Logger log = LoggerFactory.getLogger(AccessTokenRetrieverServiceImpl.class);
+
+    private ObjectMapper objectMapper = new ObjectMapper();
+
+    private ClientConfiguration clientConfiguration;
+
+    public AccessTokenRetrieverServiceImpl(ClientConfiguration clientConfiguration) {
+        Assert.notNull(clientConfiguration);
+        this.clientConfiguration = clientConfiguration;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OAuth2AccessToken getOAuth2AccessToken()
+        throws IOException, InvalidKeySpecException, NoSuchAlgorithmException {
+        String signedJwt = createSignedJWT(clientConfiguration);
+        String scope = String.join(" ", clientConfiguration.getScopes());
+
+        MediaType mediaType = MediaType.parse("application/x-www-form-urlencoded");
+        RequestBody body = RequestBody.create(mediaType, "");
+
+        OkHttpClient client = new OkHttpClient().newBuilder().build();
+
+        Request accessTokenRequest = new Request.Builder()
+            .url(clientConfiguration.getBaseUrl() +
+                "/oauth2/v1/token" +
+                "?grant_type=client_credentials" +
+                "&client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer" +
+                "&client_assertion=" + signedJwt +
+                "&scope=" + scope)
+            .method(HttpMethod.POST.name(), body)
+            .addHeader("Accept", "application/json")
+            .addHeader("Content-Type", "application/x-www-form-urlencoded")
+            .build();
+
+        Response accessTokenResponse = client.newCall(accessTokenRequest).execute();
```

Comment:
I'm not fully sure, but it appears that the returned accessToken is not stored anywhere?

If it is not currently stored, it means you are getting a token at every request.  This token should be stored in memory or cache to be used until the expire date/time.  This should retry the request if token is invalid, and get a new token before the retry.

See: https://github.com/okta/oss-technical-designs/blob/master/technical_designs/oauth-in-management-sdks.md#retry-strategy

Reply:
@bretterer I've got an option comment on that too

Reply:
@bretterer right, I discussed this with @bdemers and wanted to get this basic one (get a new token for every request) first and then build upon the token management (storage, reuse) next.  @bdemers : What did you mean by 'option comment' ? Sorry, didn't understand that.

Reply:
+1 to Retty's comment. You are supposed to reuse the same access token and retry once this has expired.

Reply:
@arvindkrishnakumar-okta sorry, there might have been some miss communication.

i was suggesting the first version:
1.) Create a token on the first request
2.) Use that token until it expires
3.) Create a new token if the current one is expired

Making an extra call to okta on every request would double the latency every request

I probably typo'd `option`, but the next version could be optimized to:
1.) Prefetch an token when the client is created (threaded)
2.) Prefetch the NEXT token a few minutes before the current expires

There are a ton of gotchas in both of those two items, so i was suggesting getting the initial code working first, and add those two if/when needed

(threading out in an existing application is tricky, we would likely need to allow for the configuration of an `ExecutorService`, and provide a reasonable default). I'm also not sure if Okta will let you use the access token after requesting another one.  If not this would still result in a blocking operation, so it might defeat the purpose

Reply:
Pushed code changes to do:

1.) Create a token on the first request
2.) Use that token until it expires
3.) Create a new token if the current one is expired

Please take a look. Working on ITs. Will push the ITs next.

Reply:
Done.